Add extensibility point for extensions to filter payment methods

ralucaStan opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR introduces a new way for extensions to register their own callbacks to determine wheater a payment method is supported.
Fixes #4621 
[Tests](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/4622), [documentation](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/4623), and the [Typescript conversion](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/4687) will be added in follow-up tickets.

### How to test the changes in this Pull Request:
1. Make sure Check payments and Cash on delivery payment methods are activated and appear in the Checkout block
2. Import the registration function somewhere in an extension
```js
import { registerPaymentMethodExtensionCallbacks} from '@woocommerce/blocks-registry';
```
3. Add some callbacks 
```js
registerPaymentMethodExtensionCallbacks( 'woocommerce-marketplace-extension',
	{
		cod: ( arg ) => { return false; },
                 // this zipcode works only for Germany
		cheque: ( arg ) => { return arg.billingData.postcode === "10473";}
	}
);
```
4. Go to the Checkout block
3. Notice that Cash on delivery is no longer available
4. Notice that Check payments are available only when the Postcode for Billing data is 10473

<!-- If you can, add the appropriate labels -->

### Changelog

> Add extensibility point for extensions to filter payment methods.

ralucaStan committed:
Add extensionsConfig when registering a payment method

The extension config has its own canMakePayment where extensions can add callback using a payment method's name.

ralucaStan committed:
Make canMakePayment a getter on PaymentMethodConfig

Because extensions can register canMakePayment callbacks for a payment method before it is registered we need to transform
canMakePayment into a getter so that it's always recalculating it's value based on the registered callbacks/

nerrad commented on assets/js/blocks-registry/payment-methods/can-make-payment-extensions.js:
```diff
@@ -0,0 +1,7 @@
+// Keeps callbacks registered by extensions for different payment methods
```
Assuming we want this config object to be something we can add additional functionality to in the future, it might be good to name the file something like `extensions-config.js` instead.

Also, maybe do this in TS from the get-go?

nerrad reviewed (commented):
Generally seems to look pretty good!

- This might be a good opportunity to convert the files to TS.
- I'd like to see some automated test coverage for the new extension logic.

I didn't test.

ralucaStan commented:
> * I'd like to see some automated test coverage for the new extension logic.

There are separate tickets for [tests](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/4622) and documentation for this logic.

ralucaStan committed:
Rename extension related config and method

opr commented on assets/js/blocks-registry/payment-methods/payment-method-config-helper.js:
```diff
@@ -8,3 +8,20 @@ export const canMakePaymentWithFeaturesCheck = ( canMakePayment, features ) => (
+	canMakePayment,
+	extensionsCallbacks
+) => ( canPayArgument ) => {
+	// validate whether the payment method is available
```
Change to use a capital letter and full stop

opr replied:
```suggestion
	// Validate whether the payment method is available.
```

opr commented on assets/js/blocks-registry/payment-methods/payment-method-config-helper.js:
```diff
@@ -8,3 +8,20 @@ export const canMakePaymentWithFeaturesCheck = ( canMakePayment, features ) => (
 	return featuresSupportRequirements && canMakePayment( canPayArgument );
 };
+
+// Filter out payment methods by callbacks registered by extensions
```
```suggestion
// Filter out payment methods by callbacks registered by extensions.
```

ralucaStan committed:
Format comments

opr commented on assets/js/blocks-registry/payment-methods/registry.js:
```diff
@@ -67,6 +67,33 @@ export const registerExpressPaymentMethod = ( options ) => {
+ * @param {string} paymentMethodName A unique string to identify the payment method client side.
+ * @param {function():any } callBack Callback defined by extensions to determine if the payment method is supported.
+ */
+export const registerPaymentMethodExtensionCallback = (
```
I am wondering what you think of changing the API to bring it more in line with the filters API. 

The API over there is like so: https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/131cb22b9c84574a88b6a4e411fabade5d1ed53c/packages/checkout/registry/index.ts#L34-L37

I was thinking we could do something similar here and change this to look like:

```js
export const registerPaymentMethodExtensionCallbacks = (
	namespace: string,
	paymentMethods: Record< string, PaymentMethodExtensionCallback >,
) 
```

Extensions would then use it like this:

```js
registerPaymentMethodExtensionCallbacks( 'test-extension'
	{
		stripe: ( arg ) => { /* code */ },
		paypal: ( arg ) => { /* code */ },
		cheque: ( arg ) => { /* code */ }
	}
);
```

ralucaStan replied:
I think it's a good suggestion for consistency reasons and also I see a great benefit in tracing back the extensions that declared those callbacks with the namespace. That would increase the complexity a bit of gathering the callbacks, but we'd also have the namespace available to track down the extensions. What do you think?

opr replied:
I was about to write the exact same comment. It's good for error reporting since we can show what extension caused the error. There is a negative tradeoff in that we would have to get the callbacks out of the namespaces in the first place, but this is only minor, and is outweighed by the positives.

ralucaStan replied:
@opr I have added the changes and re-requested review.

ralucaStan committed:
Add an extension namespace to registerPaymentMethodExtensionCallback utility

 This commit changes the API for how extensions will register their own callbacks to canMakePayment, so that they can add their namespace and also callbacks for multiple payment methods.

ralucaStan committed:
Format comments

opr commented on assets/js/blocks-registry/payment-methods/payment-method-config.js:
```diff
@@ -39,10 +42,24 @@ export default class PaymentMethodConfig {
+		this.canMakePaymentFromConfig = config.canMakePayment;
+	}
+
+	get canMakePayment() {
```
For posterity, please could you add the reason why this was moved to a getter in the PR description? We discussed this in a huddle, and I agree with your reasoning 👍🏼

opr commented on assets/js/blocks-registry/payment-methods/payment-method-config.js:
```diff
@@ -39,10 +42,24 @@ export default class PaymentMethodConfig {
+			this.canMakePaymentFromConfig,
 			this.supports.features
 		);
+
```
I think we'd benefit from describing the purpose of the next statement. You can figure it out by reading, but this makes it a bit quicker :)

```suggestion
// Loop through all calbacks to check if there are any registered for this payment method.
```

opr commented on assets/js/blocks-registry/payment-methods/registry.js:
```diff
@@ -67,6 +67,44 @@ export const registerExpressPaymentMethod = ( options ) => {
+		);
+
+		throw new Error(
+			`All callbacks provided to registerPaymentMethodExtensionCallbacks must be functions. The callbacks for :${ nonFunctionCallbacks.join(
```
```suggestion
			`All callbacks provided to registerPaymentMethodExtensionCallbacks must be functions. The callbacks for: ${ nonFunctionCallbacks.join(
```

The formatting is slightly off here
![image](https://user-images.githubusercontent.com/5656702/132357165-b2d59c3b-4ca5-4a10-b2df-9b7a08f7b10d.png)

opr commented on assets/js/blocks-registry/payment-methods/registry.js:
```diff
@@ -67,6 +67,44 @@ export const registerExpressPaymentMethod = ( options ) => {
+		);
+	}
+
+	// Set namespace up as an empty object
```
```suggestion
	// Set namespace up as an empty object.
```

opr commented on assets/js/blocks-registry/payment-methods/registry.js:
```diff
@@ -67,6 +67,44 @@ export const registerExpressPaymentMethod = ( options ) => {
+			( key ) => typeof callbacks[ key ] !== 'function'
+		);
+
+		throw new Error(
```
Currently, if one extension registers a bad callback, it stops all subsequent extensions from registering theirs.

Instead of throwing an error here, let's just `console.error` and return early from the function?

I don't think there's anywhere further up the stack that we'd benefit from catching the error.

You can simulate this by adding another call to `registerPaymentMethodExtensionCallbacks` immediately after the initial one, ensuring one of the callbacks is not a function.

```js
registerPaymentMethodExtensionCallbacks( 'oprs-marketplace-extension',
		{
			cod: ( arg ) => { return false; },
			stripe: null,
		}
	);
```

ralucaStan replied:
@opr 👏 for spotting this. 
I am thinking we should not return early, just use `console.error` and proceed to the other callbacks the extension might have. So if one callback is not a function we are registering the ones that actually are. Like so:
```js
export const registerPaymentMethodExtensionCallbacks = (
	namespace,
	callbacks
) => {
	// Set namespace up as an empty object.
	canMakePaymentExtensionsCallbacks[ namespace ] = {};

	Object.entries( callbacks ).forEach(
		( [ paymentMethodName, callback ] ) => {
			if ( typeof callback === 'function' ) {
				canMakePaymentExtensionsCallbacks[ namespace ][
					paymentMethodName
				] = callback;
			} else {
				console.error(
					`All callbacks provided to registerPaymentMethodExtensionCallbacks must be functions. The callback for the ${ paymentMethodName } payment method was not a function. Error occurred when trying to register callbacks for the ${ namespace } namespace.`
				);
			}
		}
	);
};
```

But I do think we should return early if the namespace is already used. So instead of letting a call of registerPaymentMethodExtensionCallbacks that overwrites existing callbacks under the same namespace, just console.error and return. What do you think?

opr replied:
> I am thinking we should not return early

OK :)

> But I do think we should return early if the namespace is already used. So instead of letting a call of registerPaymentMethodExtensionCallbacks that overwrites existing callbacks under the same namespace, just console.error and return. What do you think?

Great idea!

opr commented on assets/js/blocks-registry/payment-methods/payment-method-config-helper.js:
```diff
@@ -8,3 +8,41 @@ export const canMakePaymentWithFeaturesCheck = ( canMakePayment, features ) => (
+					err
+				);
+				// every expects a return value at the end of every arrow function and
+				// this ensures that the error is ignored when computing the whole result
```
```suggestion
				// this ensures that the error is ignored when computing the whole result.
```

opr commented on assets/js/blocks-registry/payment-methods/payment-method-config-helper.js:
```diff
@@ -8,3 +8,41 @@ export const canMakePaymentWithFeaturesCheck = ( canMakePayment, features ) => (
+	extensionsCallbacks,
+	paymentMethodName
+) => ( canPayArgument ) => {
+	// Validate whether the payment method is available.
```
```suggestion
	// Validate whether the payment method is available based on its own criteria first.
```

opr reviewed (changes requested):
This is fantastic Raluca 👏🏼  I have added a few (admittedly pedantic) comments about comment punctuation. 

I tested with:

- Two different extensions adding valid callbacks. ✅ 
- Two different extensions adding valid callbacks, one returning true for a payment metod, and one returning false for the same, I expected NOT to see the payment method ✅ 
- An extension registering a callback that is not a function - An error is thrown (see comments for improvements here) - If two extensions try to register callbacks, but the first one tries registering a callback that isn't a function, then the second one doesn't get chance to register. ❌ 
- If a callback that is correctly registered, but for some reason is causing an error, other extensions' callbacks can still run ✅ 

For reference, here are the two registrations I'm using:

```js
	registerPaymentMethodExtensionCallbacks( 'woocommerce-marketplace-extension',
		{
			cod: ( arg ) => { return false; },
                         // A callback that is not a function. This should cause an error, but other extensions should still be allowed to register.
                         cheque: null,
			stripe: () => { aaa.bbb = 1; return true; }
		}
	);

	registerPaymentMethodExtensionCallbacks( 'oprs-marketplace-extension',
		{
                         // COD should not show because the first callback returns false.
			cod: ( arg ) => { return true; },
                         // Because the stripe callback above causes an error, it should be skipped. Stripe should only show when postcode is 10437 (in Germany).
			stripe: ( arg ) => arg.billingData.postcode === '10473',
                         // Because cheque registration should have failed above, we should still be allowed to see cheque in the payment method list.
			cheque: () => { return true },
		}
	);
```

ralucaStan committed:
Update assets/js/blocks-registry/payment-methods/payment-method-config.js

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

ralucaStan committed:
Update assets/js/blocks-registry/payment-methods/registry.js

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

ralucaStan committed:
Update assets/js/blocks-registry/payment-methods/payment-method-config-helper.js

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

ralucaStan committed:
Update assets/js/blocks-registry/payment-methods/payment-method-config-helper.js

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

ralucaStan committed:
Update assets/js/blocks-registry/payment-methods/registry.js

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

ralucaStan committed:
Fix eslint warning

ralucaStan committed:
Handle errors at registerPaymentMethodExtensionCallbacks level

opr commented on assets/js/blocks-registry/payment-methods/registry.js:
```diff
@@ -67,6 +67,42 @@ export const registerExpressPaymentMethod = ( options ) => {
+				} else {
+					// eslint-disable-next-line no-console
+					console.error(
+						`All callbacks provided to registerPaymentMethodExtensionCallbacks must be functions. The callback for the ${ paymentMethodName } payment method was not a function. Error occurred when trying to register callbacks for the ${ namespace } namespace.`
```
```suggestion
						`All callbacks provided to registerPaymentMethodExtensionCallbacks must be functions. The callback for the ${ paymentMethodName } payment method in the ${ namespace} namespace was not a function.`
```

opr reviewed (approved):
This is looking great @ralucaStan - I have made one final suggestion about the error message when registering non-function callbacks. After looking at it for some time it felt a little clunky. My suggestion makes it a bit easier to read imo.

Approving because this change is non-blocking and the rest of the PR is working fine now 🥳

opr reviewed (commented):
Actually, before we merge, I am wondering about the name of `registerPaymentMethodExtensionCallbacks` why isn't this prefixed with `__experimental`?

Should it be renamed before merging?

nerrad commented:
> Actually, before we merge, I am wondering about the name of registerPaymentMethodExtensionCallbacks why isn't this prefixed with __experimental?

In this particular case I think I'm okay with skipping experimental for this interface. It's fairly straightforward and I think other alternatives might have more riskiness/uncertainty attached to them. This is just an opinion though - so I'm happy to defer to others here.

ralucaStan committed:
Update assets/js/blocks-registry/payment-methods/registry.js

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

ralucaStan commented:
> Actually, before we merge, I am wondering about the name of `registerPaymentMethodExtensionCallbacks` why isn't this prefixed with `__experimental`?
> 
> Should it be renamed before merging?

I also thought about this, but after looking at the history of the registry file I saw that the exports didn't have an experimental prefix.

opr committed:
Fix formatting issues
